pkg/geo/geomfn/subdivide.go

```diff
@@ -0,0 +1,364 @@
+		for i := 0; i < gt.NumGeoms(); i++ {
+			err := fn(gt.Geom(i))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	case *geom.MultiPolygon:
+		for i := 0; i < gt.NumPolygons(); i++ {
+			err := fn(gt.Polygon(i))
+			if err != nil {
+				return err
+			}
+		}
+	case *geom.MultiLineString:
+		for i := 0; i < gt.NumLineStrings(); i++ {
+			err := fn(gt.LineString(i))
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func calculateSplitPoint(
+	gt geom.T, nVertices int, splitHorizontally bool, clip *geo.CartesianBoundingBox,
+) (float64, error) {
+	var pivot float64
+	switch gt := gt.(type) {
+	case *geom.Polygon:
+		var err error
+		pivot, err = findPolygonsMostCentralPointValue(gt, nVertices, splitHorizontally, clip)
+		if err != nil {
+			return 0, errors.Newf("error finding most central point for polygon: %v", err)
+		}
+		if splitHorizontally {
+			// Ignore point on the boundaries
+			if clip.LoX == pivot || clip.HiX == pivot {
+				pivot = (clip.LoX + clip.HiX) / 2
+			}
+		} else {
+			// Ignore point on the boundaries
+			if clip.LoY == pivot || clip.HiY == pivot {
+				pivot = (clip.LoY + clip.HiY) / 2
+			}
+		}
+	default:
+		if splitHorizontally {
+			pivot = (clip.LoX + clip.HiX) / 2
+		} else {
+			pivot = (clip.LoY + clip.HiY) / 2
+		}
+	}
+	return pivot, nil
+}
+
+// findPolygonsMostCentralPointValue finds the value of the most central point for a geom.Polygon.
+// Can be X or Y value, depending on the splitHorizontally argument.
+func findPolygonsMostCentralPointValue(
+	poly *geom.Polygon, nVertices int, splitHorizontally bool, clip *geo.CartesianBoundingBox,
+) (float64, error) {
+	pivot := math.MaxFloat64
+	ringToTrimIndex := 0
+	ringArea := float64(0)
+	pivotEps := math.MaxFloat64
+	var ptEps float64
+	nVerticesOuterRing, err := countVertices(poly.LinearRing(0))
+	if err != nil {
+		return 0, errors.Newf("error counting vertices: %v", err)
+	}
+	// if there are more points in holes than in outer ring
+	if nVertices >= 2*nVerticesOuterRing {
+		// trim holes starting from biggest
+		for i := 1; i < poly.NumLinearRings(); i++ {
+			currentRingArea := poly.LinearRing(i).Area()
+			if currentRingArea >= ringArea {
+				ringArea = currentRingArea
+				ringToTrimIndex = i
+			}
+		}
+	}
+	pa := poly.LinearRing(ringToTrimIndex)
+	// Find the most central point in the chosen ring
+	for j := 0; j < pa.NumCoords(); j++ {
+		var pt float64
+		if splitHorizontally {
+			pt = pa.Coord(j).X()
+			xHalf := (clip.LoX + clip.HiX) / 2
+			ptEps = math.Abs(pt - xHalf)
+		} else {
+			pt = pa.Coord(j).Y()
+			yHalf := (clip.LoY + clip.HiY) / 2
+			ptEps = math.Abs(pt - yHalf)
+		}
+		if ptEps < pivotEps {
+			pivot = pt
+			pivotEps = ptEps
+		}
+	}
+	return pivot, nil
+}
+
+// clipGeomTByBoundingBox returns a result of intersection and simplification of the geom.T and bounding box provided.
+// The result is also a geom.T.
+func clipGeomTByBoundingBox(gt geom.T, clip *geo.CartesianBoundingBox) (geom.T, error) {
```

Comment:
nit: `clipGeomTByBoundingBoxForSubdivide` (basically, these specialised functions for subdivide sound like they work for other stuff, but I think there's special logic here such that it is not)

Reply:
Makes sense, fixed